src/daft-session/src/session.rs

```diff
@@ -28,17 +29,82 @@ struct SessionState {
     catalogs: Bindings<CatalogRef>,
     /// Bindings for the attached tables.
     tables: Bindings<TableRef>,
-    // TODO execution context
+    /// User defined functions
+    functions: Bindings<WrappedUDFClass>,
 }
 
 impl Session {
+    /// Creates an independent copy of the Session with its own separate state.
+    ///
+    /// This method performs a complete clone of the underlying `SessionState`,
+    /// creating a new `Arc<RwLock<SessionState>>` that contains the cloned state.
+    /// Any changes made to the state of the forked session will not affect the
+    /// original session and vice versa.
+    ///
+    /// # Comparison with `clone_ref()`
+    ///
+    /// Unlike `clone_ref()` which only creates a new reference to the same underlying state,
+    /// `fork()` creates a complete copy of the state. This makes `fork()` more expensive
+    /// but ensures complete isolation between the original and new session.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// let original_session = Session::empty();
+    /// // ... configure original_session ...
+    ///
+    /// // Create an independent copy with the same initial state
+    /// let forked_session = original_session.fork();
+    ///
+    /// // Changes to forked_session won't affect original_session
+    /// ```
+    ///
+    pub fn fork(&self) -> Self {
+        let state = self.state().clone();
+        Self {
+            state: Arc::new(RwLock::new(state)),
+        }
+    }
```

Comment:
This is interesting. Does this mean every call to `daft.sql` will fork the session?

Reply:
yes. Am planning on refactoring the `SQLContext` and Session in a subsequent PR. This is just a stopgap until that refactor is complete